### PR TITLE
Implement auto NUTS2 version detector for UK

### LINKED ISSUE #122: Auto-NUTS2 version for UK
Write a function to automatically detect which NUTS boundary year should be used for each year within an indicator dataset.

This should be able to take a dataframe with columns `nuts_id` and `year` and create a new field for `nuts_year`.

This is to account for the fact that there is a 2 year lag between new NUTS standards being introduced and their enforcement. We don't currently know when different organisations choose to self-enforce the new boundaries within this lag period.

### PR DESCRIPTION
**Pull request by georgerichardson**
Closes #122 

```
from beis_indicators.utils.nuts_utils import auto_nuts

# create an indicator dataframe
df = pd.DataFrame({
    'value': values # some values
    'nuts_id': nuts_ids, # the NUTS regions corresponding to the values
    'year': years, # the years corresponding to the values
})

df = auto_nuts(df, 'year', 'nuts_id') # infers the NUTS versions in the data and auto generates the nuts_year_spec column

```

Each version of NUTS has a unique set of regions. Every year, some region codes are dropped and new ones are introduced.

In the situation where we have indicator values for regions and the year that the indicator refers to, but where no NUTS version information is provided, `auto_nuts` can be used to infer the NUTS boundary version.

The function uses containment as a measure to find the best match between the region codes provided and the region codes for each NUTS version. If no perfect match is found, then it uses the highest containment score. If there is a tie, then the earliest valid version is used (i.e. if the indicator values correspond to the year 2013, then the only NUTS versions that are considered are those which were enforced on or after 2013).

### CODE REVIEW

File: beis_indicators/utils/nuts_utils.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+NUTS_ENFORCED = {
+    2003: 2003,
+    2006: 2008,
+    2010: 2012,
+    2013: 2015,
+    2016: 2018
+}
+
+NUTS_IDS ={
+        2003: {'UKC2', 'UKK4', 'UKL2', 'UKG3', 'UKK2', 'UKD1', 'UKF2', 'UKG1',
```
Comment (mindrones) [inline beis_indicators/utils/nuts_utils.py:16]: By looking at this dict, I would rename this utilities to point out that they work on UK and NUTS2 only and I'd edit the title of both the issue and the PR.

Here I'd use `NUTS2_UK_IDS`.

File: beis_indicators/utils/nuts_utils.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    '''nuts_earliest
+    Returns the earliest possible NUTS version for a year
+    based on the enforcement date.
+    '''
+    for k, v in NUTS_ENFORCED.items():
+        if year >= v:
+            earliest = k
+    return earliest
+
+def _detect_nuts(ids, year):
```
Comment (mindrones) [inline beis_indicators/utils/nuts_utils.py]: here I'd use `_detect_nuts2_uk`

File: beis_indicators/utils/nuts_utils.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        return year_inferred
+    
+    # if there is not a perfect match then chose the earliest and best
+    # possible match
+    elif np.sum(perfect) != 1:
+        best = np.argwhere(containments == np.max(containments)).ravel()
+        year_inferred = years[best[0]]
+        return year_inferred
+
+def auto_nuts(df, year='year', nuts_id='nuts_id'):
```
Comment (mindrones) [inline beis_indicators/utils/nuts_utils.py]: here I'd use `auto_nuts2_uk`

**Review by mindrones — CHANGES_REQUESTED** [review]
renaming comments

File: beis_indicators/utils/nuts_utils.py (hunk 1 above)
Comment (georgerichardson) [reply beis_indicators/utils/nuts_utils.py:16]: 👍

File: beis_indicators/utils/nuts_utils.py (hunk 2 above)
Comment (georgerichardson) [reply beis_indicators/utils/nuts_utils.py]: Ok. I think long term these will become generic functions though

Comment (mindrones) [reply beis_indicators/utils/nuts_utils.py]: Long term, right now they aren't, so we should use naming that is appropriate for this version of the source code.

**Review by mindrones — APPROVED** [review]

File: beis_indicators/utils/nuts_utils.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            'UKM2', 'UKE4', 'UKE1', 'UKK1', 'UKE3', 'UKI5', 'UKK3', 'UKG2', 
+            'UKM5', 'UKL1', 'UKI4', 'UKD7', 'UKJ2', 'UKN0', 'UKH1', 'UKM3', 
+            'UKH2', 'UKF3', 'UKD4', 'UKM6', 'UKI3', 'UKC1', 'UKD6', 'UKF1'}, 
+        2016: {'UKM7', 'UKC2', 'UKM8', 'UKK4', 'UKL2', 'UKG3', 'UKK2', 'UKI6', 
+            'UKD1', 'UKF2', 'UKI7', 'UKG1', 'UKD3', 'UKH3', 'UKJ4', 'UKE2', 
+            'UKJ1', 'UKJ3', 'UKE4', 'UKE1', 'UKK1', 'UKE3', 'UKI5', 'UKM9', 
+            'UKK3', 'UKG2', 'UKM5', 'UKL1', 'UKI4', 'UKD7', 'UKJ2', 'UKN0', 
+            'UKH1', 'UKF3', 'UKH2', 'UKM6', 'UKD4', 'UKC1', 'UKI3', 'UKD6', 'UKF1'}}
+
+NUTS_YEARS = np.array(list(NUTS_IDS.keys()))
```
Comment (Juan-Mateos) [inline beis_indicators/utils/nuts_utils.py]: This should be NUTS2_UK_IDS as above

File: beis_indicators/utils/nuts_utils.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+def _year_containments(ids, years):
+    containments = []
+    for year in years:
+        year_ids = NUTS_IDS[year]
```
Comment (Juan-Mateos) [inline beis_indicators/utils/nuts_utils.py]: NUTS2_UK_IDS

**Review by Juan-Mateos — CHANGES_REQUESTED** [review]
Need to change some arg names inside the functions (NUTS_IDS to NUTS2_UK_IDS)

**Comment by georgerichardson** [comment]
Thanks for spotting those @Juan-Mateos 

Should be ok to approve now

**Review by mindrones — APPROVED** [review]

**Pull request merged**